### PR TITLE
[WIP] Set the Content-Type of http.Header when the request body type is string/[]byte/io.Reader

### PR DESCRIPTION
**Pull request by sxllwx**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

When using a **non-runtime.Object** object as the http body, the Content-Type in http.Header is not set correctly

#### Which issue(s) this PR fixes:

Fixes # https://github.com/kubernetes/client-go/issues/980

#### Special notes for your reviewer:

Deleted the code here https://github.com/kubernetes/kubernetes/blob/f7cff077d5693d2c5e199aa18c002a4c25667f4c/staging/src/k8s.io/client-go/rest/request.go#L457. Set the Content-type at the time of NewRequest uniformly

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sxllwx** [comment]
@roycaihw PTAL😄

File: staging/src/k8s.io/client-go/rest/request.go [hunk 1]
```diff
@@ -151,6 +151,7 @@ func NewRequest(c *RESTClient) *Request {
 		r.SetHeader("Accept", c.content.AcceptContentTypes)
 	case len(c.content.ContentType) > 0:
 		r.SetHeader("Accept", c.content.ContentType+", */*")
+		r.SetHeader("Content-Type", r.c.content.ContentType)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/request.go:154]: setting Content-Type when we don't know if we're even sending a body in the request is not correct

File: staging/src/k8s.io/client-go/rest/request.go [hunk 2]
```diff
@@ -454,7 +455,6 @@ func (r *Request) Body(obj interface{}) *Request {
 		}
 		glogBody("Request Body", data)
 		r.body = bytes.NewReader(data)
-		r.SetHeader("Content-Type", r.c.content.ContentType)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/request.go:457]: restore this, but only do it if `len(r.c.content.ContentType) > 0`?

File: staging/src/k8s.io/client-go/rest/request.go [hunk 3]
```diff
@@ -454,7 +455,6 @@ func (r *Request) Body(obj interface{}) *Request {
 		}
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/request.go:455]: for the string/[]byte/reader cases above, I could see setting Content-Type if there's no existing content-type header and len(r.c.content.ContentType) > 0

**Comment by sxllwx** [comment]
@liggitt  Thank you for your review. According to your suggestions, I have made some changes. PTAL

File: staging/src/k8s.io/client-go/rest/request.go [hunk 4]
```diff
@@ -432,11 +432,20 @@ func (r *Request) Body(obj interface{}) *Request {
...
 		glogBody("Request Body", t)
 		r.body = bytes.NewReader(t)
+		if len(r.headers.Get("Content-Type")) == 0 && len(r.c.content.ContentType) > 0{
+			r.SetHeader("Content-Type", r.c.content.ContentType)
+		}
 	case io.Reader:
 		r.body = t
+		if len(r.headers.Get("Content-Type")) == 0 && len(r.c.content.ContentType) > 0{
+			r.SetHeader("Content-Type", r.c.content.ContentType)
+		}
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/request.go:448]: @smarterclayton I think this is a reasonable default behavior, and clients that want different behavior can explicitly call `r.SetHeader("Content-Type", ...)` to get the behavior they want (before or after calling `Body`). What do you think?

**Comment by liggitt** [comment]
the test failures look relevant:

```
E0615 05:07:30.660026   58063 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"provided data does not appear to be a protobuf message, expected prefix [107 56 115 0]"}: provided data does not appear to be a protobuf message, expected prefix [107 56 115 0]
```

would like @smarterclayton's feedback on the approach as well

**Comment by sxllwx** [comment]
After tracing the error. It is found that the kube-apiserver side uses the Protobuf decoder to send the body encoded in JSON to the client. kube-controller-manager selects protobuf as the serialization method with kube-apiserver by default.

---

The namespace-controller in kube-controller-manager will be used when executing [deleteCollection](https://github.com/kubernetes/kubernetes/blob/9c4905e7af0d070d139c06638ab9868fd3521883/pkg/controller/namespace/deletion/namespaced_resources_deleter.go#L320) `deleteOptionsCodec` encodes `DeleteOptions` ([specific code snippet](https://github.com/kubernetes/kubernetes/blob/9c4905e7af0d070d139c06638ab9868fd3521883/staging/src/k8s.io/client-go/metadata/metadata.go# L142)), and the Json serialization method is used here.

> @liggitt Now I feel that we should be cautious about making changes to request.go. This problem can be fixed by explicitly adding the setting of Content-Type in the process of constructing Request in `dynamicResourceClient`. What do you think?

**Comment by liggitt** [comment]
I think that's a bug in the metadata client... I *think* it should be calling Body(&opts) rather than encoding to JSON in conflict with its configured content-type

**Comment by sxllwx** [comment]
@liggitt 

Correct Me If I'm Wrong. During the test, it was found that the current kube-apiserver cannot complete Protobuf decoding metav1.DeleteOptions. The specific error is:

> E0714 03:34:39.283466 57205 status.go:71] apiserver received an error that is not an metav1.Status: &runtime.notRegisteredErr{schemeName:"k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:30 ", gvk:schema.GroupVersionKind{Group:"meta.k8s.io", Version:"__internal", Kind:"DeleteOptions"}, target:runtime.GroupVersioner(nil), t:reflect.Type(nil)}: no kind "DeleteOptions" is registered for the internal version of group "meta.k8s.io" in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:30"

After investigation, it was found that the `legacyscheme` did not register metav1.DeleteOptions. Should I add the installer of the corresponding scheme in the controllerplane?

**Comment by liggitt** [comment]
> After investigation, it was found that the `legacyscheme` did not register metav1.DeleteOptions. Should I add the installer of the corresponding scheme in the controllerplane?

hmm... maybe. if there are server-side changes required to support this, we'll need to make/test/release those first, likely multiple releases before making client-side changes.

**Comment by sxllwx** [comment]
> support

I am sorry that I have not responded to you in time due to heavy workload. I also agree to split this modification into two sides: Client and Server. I will continue to make/test/releas for this issue, thanks for your suggestions. 😄

**Comment by sxllwx** [comment]
> > After investigation, it was found that the `legacyscheme` did not register metav1.DeleteOptions. Should I add the installer of the corresponding scheme in the controllerplane?
> 
> hmm... maybe. if there are server-side changes required to support this, we'll need to make/test/release those first, likely multiple releases before making client-side changes.

I am sorry that I have not responded to you in time due to heavy workload. I also agree to split this modification into two sides: Client and Server. I will continue to make/test/releas for this issue, thanks for your suggestions. 😄

**Pull request closed without merging**